tests/test_uri.py

```diff
@@ -19,474 +19,32 @@
-
-        # Grouping but check that single files are returned in a single group
-        # at the end
-        file2 = test_dir_uri.join("config/templates/templates-bad.yaml")
-        found = list(
-            ButlerURI.findFileResources([file, file2, test_dir_uri.join("config/dbAuth")], grouped=True)
-        )
-        self.assertEqual(len(found), 2)
-        self.assertEqual(list(found[1]), [file, file2])
-
-        with self.assertRaises(ValueError):
-            list(file.walk())
-
-    def testRootURI(self):
-        """Test ButlerURI.root_uri()."""
-        uri = ButlerURI("https://www.notexist.com:8080/file/test")
-        uri2 = ButlerURI("s3://www.notexist.com/file/test")
-        self.assertEqual(uri.root_uri().geturl(), "https://www.notexist.com:8080/")
-        self.assertEqual(uri2.root_uri().geturl(), "s3://www.notexist.com/")
-
-    def testJoin(self):
-        """Test .join method."""
-
-        root_str = "s3://bucket/hsc/payload/"
-        root = ButlerURI(root_str)
-
-        self.assertEqual(root.join("b/test.txt").geturl(), f"{root_str}b/test.txt")
-        add_dir = root.join("b/c/d/")
-        self.assertTrue(add_dir.isdir())
-        self.assertEqual(add_dir.geturl(), f"{root_str}b/c/d/")
-
-        quote_example = "b&c.t@x#t"
-        needs_quote = root.join(quote_example)
-        self.assertEqual(needs_quote.unquoted_path, f"/hsc/payload/{quote_example}")
-
-        other = ButlerURI("file://localhost/test.txt")
-        self.assertEqual(root.join(other), other)
-        self.assertEqual(other.join("b/new.txt").geturl(), "file://localhost/b/new.txt")
-
-        joined = ButlerURI("s3://bucket/hsc/payload/").join(ButlerURI("test.qgraph", forceAbsolute=False))
-        self.assertEqual(joined, ButlerURI("s3://bucket/hsc/payload/test.qgraph"))
-
-        with self.assertRaises(ValueError):
-            ButlerURI("s3://bucket/hsc/payload/").join(ButlerURI("test.qgraph"))
-
-    def testTemporary(self):
-        with ButlerURI.temporary_uri(suffix=".json") as tmp:
-            self.assertEqual(tmp.getExtension(), ".json", f"uri: {tmp}")
-            self.assertTrue(tmp.isabs(), f"uri: {tmp}")
-            self.assertFalse(tmp.exists(), f"uri: {tmp}")
-            tmp.write(b"abcd")
-            self.assertTrue(tmp.exists(), f"uri: {tmp}")
-            self.assertTrue(tmp.isTemporary)
-        self.assertFalse(tmp.exists(), f"uri: {tmp}")
-
-        tmpdir = ButlerURI(self.tmpdir, forceDirectory=True)
-        with ButlerURI.temporary_uri(prefix=tmpdir, suffix=".yaml") as tmp:
-            # Use a specified tmpdir and check it is okay for the file
-            # to not be created.
-            self.assertFalse(tmp.exists(), f"uri: {tmp}")
-        self.assertTrue(tmpdir.exists(), f"uri: {tmpdir} still exists")
+    def test_uri(self):
+        netloc = "netloc"
+        path = "/a/b/c.txt"
+        for scheme in ("file", "s3", "https"):
+            uri_str = f"{scheme}://{netloc}{path}"
```

Comment:
Just a little nitpick. Only `scheme` varies here and `netloc` and `path` are not used anywhere else. So I'm not sure if defining them is worth the trouble. On the other hand having them doesn't hurt much either so if you have reasons to keep them I won't object.

Reply:
Good point. I changed my mind half way through.